### ISSUE TITLE
CL caching for continuous integration

### ISSUE THREAD

**Issue by kg** [sq/JSIL#833]
AppVeyor has an option to cache folders between builds; if we tweak the test harness to allow setting a location for the compiled test temp directory, we could cache it and cut out the ~hundreds of ms per test spent running the C# compiler.

csc is really slow on mono, too, so the same hack would help to make travis CI faster.

Thoughts? I don't think the test cache has ever caused problems for me, but maybe you can think of a reason not to do this, @iskiselev

**Comment by kg** [comment]
I overhauled the compile caching system (so that it would work even if the repository is blown away) and enabled it on appveyor, and it reduces the time spent running tests by around 30 seconds. So that works! It'd be nice if we could do this on Travis, but their documentation suggests it's only supported for private repositories...

**Comment by iskiselev** [comment]
Yes, on Travis it is supported only for private repositories, which really should be read as "only for paid users" :)
If we have some storage, we could probably pack them and upload it as last build step - and download them as first step.

**Comment by kg** [comment]
I could probably reserve some storage for it on my linode if we figure out a secure way to do it.

**Comment by iskiselev** [comment]
You could easily do it for your own builds - you can add secured environment variables with keys to your storage, but this variables will be unavailable for any PR build not from sq/JSIL repo (to not compromise them).
I've thought about registering DropBox account to store such artifacts (we may also want upload archive with all generated scripts for simpler debug).
As this DropBox account will not have any sensitive content, we may include access key inside .yml file (or better inside travis-ci ui).
But in this case you should not use secured environment variables on Travis-CI, as really anybody will be able change our artifacts on DropBox.

**Comment by kg** [comment]
According to travis support you can actually use their caching on open source projects as long as they use containers. Did you set us up with containers? It seems like you did.

**Comment by kg** [comment]
Hacked together a basic caching system for travis CI. It stores nuget packages and the test compilation cache onto my linode. There's some room for improvement (the upload/download take some time) but from my initial test it shaves a few hundred seconds off test runtime, and the nuget package caching saves 15+ seconds on its own.

**Comment by iskiselev** [comment]
I've not tried cache there yet as I've supposed it cpuld be used only in
paid accounts. If support state different we may try, as we use their
container infrastructure.

**Comment by iskiselev** [comment]
Really I've just looked on logs - and right now I don't see any benefits from caching in compilation time right now. It haven't saved any time. Maybe it was some local problem - so let's look on total build time for next builds.

**Comment by kg** [comment]
It definitely removes the ~500ms compiles, the problem is that Travis instances are randomly 2-5x slower (you can see this via translation time variances). I assume the machines they use run multiple builds at once.

On August 31, 2015 10:23:53 AM PDT, Igor Kiselev <notifications@github.com> wrote:
>Really I've just looked on logs - and right now I don't see any
>benefits from caching in compilation time right now. It haven't saved
>any time. Maybe it was some local problem - so let's look on total
>build time for next builds.
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/sq/JSIL/issues/833#issuecomment-136434498

- kg (mobile)

**Comment by iskiselev** [comment]
@kg, what for are you using $JSIL_STORAGE_TAG in your upload/download script? Is it secured parameter?

**Comment by kg** [comment]
There are separate storage buckets for the various bits of data and individual build machines. There's a single secured parameter that's used as an authentication key, also.

**Comment by kg** [comment]
See here: https://github.com/sq/JSIL/blob/master/.travis.yml#L28

**Comment by iskiselev** [comment]
Thank you, get it!
I'm working now on publishing artifacts of each build (to be able to debug it).
My current plan:
1. Use [Yandex.Disk](https://disk.yandex.com/invite/?hash=OWD2151Q) as storage - as they has WebDav API
2. Share artifacts folder (ReadOnly) with special created account.
3. If user has defined his own OAuth key - use it for both uploading artifacts/cache and downloading cache.
4. Otherwise use user (2) OAuth key for downloading cache.

After it, cache will be resolved for each build, even pull request.

**Comment by kg** [comment]
That's a good idea. I thought about publishing build artifacts for debugging but couldn't find an easy way to do it on travis or appveyor.

**Comment by iskiselev** [comment]
@kg, how much time does it take to download nuget packages from your storage? Or better, what network speed do you see?
With Yandex it gives me about 350KB per second which is not fast :(

**Comment by kg** [comment]
Nuget restore off my linode is typically ~1 second, 2 seconds at most. It's
a linode in Fremont with good bandwidth.

If necessary I can tweak it so that the same mechanism can be used to push
artifacts into a httpd-accessible directory. I'll still want to
authenticate it with a hidden secret, though.

**Comment by iskiselev** [comment]
Looks like Travis caching is working.
So, in that case I will use it for packages/compiled tests and Yandex.Disk for artifacts and translate tests.
